### PR TITLE
Modify library structure to support web components

### LINKED ISSUE #1315: [Enhancement] Web Components: Create new lib structure for StencilJS
**Is your enhancement request related to a problem? Please describe.**
- Enable the use of Kirby in other frameworks than Angular (and vanilla js projects)

**Describe the solution you'd like**
[ ] Create new `@kirbydesign/core` lib using StencilJS

**Additional context**
See existing POC branch: https://github.com/kirbydesign/designsystem/tree/poc/stenciljs
Also see: https://stenciljs.com/docs/angular

### PR DESCRIPTION
**Pull request by flemcito**
This PR closes #1315 

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

A major restructuring of the repo. 
- Introducing the @kirbydesign/core package. 

- [ ] Bugfix
- [ ] Feature
- [x] Enhancement (to existing content)
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [x] CI related changes
- [x] Documentation content changes
- [x] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## Which issue documents the current behaviour?

Issue Number: #[replace with issue number]

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by RasmusKjeldgaard** [comment]
TODO:
[] Create separate branch for autogenerating mock based on this branch, and merge to this
[] **!IMPORTANT!** Before merge, make sure to copy over the content of the global styles that have been moved, so they match the current state of master

File: libs/designsystem/src/lib/helpers/color-helper.scss [hunk 1]
```diff
@@ -1,4 +1,4 @@
 @use 'sass-to-ts';
-@use '../scss/themes/colors';
+@use '../../../../core/src/scss/themes/colors';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/helpers/color-helper.scss:2]: use `~@kirbydesign/core` alias

File: libs/designsystem/src/lib/helpers/design-token-helper.scss [hunk 2]
```diff
@@ -1,5 +1,5 @@
 @use 'sass-to-ts';
-@use '../scss/base/variables';
+@use '../../../../core/src/scss/base/variables';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/helpers/design-token-helper.scss:2]: use `~@kirbydesign/core` alias

File: libs/designsystem/src/lib/scss/_utils.scss [hunk 3]
```diff
@@ -1,8 +1,2 @@
-/*  
-    General imports for components
-    NOTICE: Must not contain anything that generates classes or else it will be generated for every component that imports this file.
-*/
-@import './base/functions';
-@import './base/variables';
-@import './base/include-media';
-@import './themes/colors';
+// TODO: Remove this file when all components are moved to core
+@import '../../../../core/src/scss/utils';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/scss/_utils.scss:2]: use ~@kirbydesign/core alias

File: libs/core/stencil.config.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  plugins: [sass({ injectGlobalPaths: ['./src/scss/utils'] })],
+  outputTargets: [
+    angularOutputTarget({
+      componentCorePackage: '@kirbydesign/core',
+      directivesProxyFile: '../designsystem/src/lib/components/proxies.ts',
+    }),
+    {
+      type: 'dist',
+      esmLoaderPath: '../loader',
+      // copy: [{ src: './scss', dest: '../scss' }],
```
Comment (RasmusKjeldgaard) [inline libs/core/stencil.config.ts:16]: Remove comment

File: libs/designsystem/tslint.json [hunk 5]
```diff
@@ -3,5 +3,8 @@
   "rules": {
     "directive-selector": [true, "attribute", "kirby", "camelCase"],
     "component-selector": [true, "element", "kirby", "kebab-case"]
+  },
+  "linterOptions": {
+    "exclude": ["src/lib/components/angular-component-lib/**.*"]
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/tslint.json:8]: remember to change this with new /stencil/ structure

File: package.json [hunk 6]
```diff
@@ -160,7 +175,7 @@
   "husky": {
     "hooks": {
       "pre-commit": "npm run sync-node-version && npm run generate-mocks && lint-staged",
-      "pre-push": "npm run test:headless-browser:single",
+      "pre-push": "npm run test:core:single && npm run test:headless-browser:single",
```
Comment (RasmusKjeldgaard) [inline package.json:178]: extract these two run steps into `test:libs` or similar

File: tsconfig.base.json [hunk 7]
```diff
@@ -8,9 +8,10 @@
     "moduleResolution": "node",
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
-    "target": "es2015",
+    "allowUnreachableCode": false,
+    "target": "es2017",
```
Comment (RasmusKjeldgaard) [inline tsconfig.base.json:12]: why is this needed? Check up on consequences.

File: tsconfig.base.json [hunk 8]
```diff
@@ -8,9 +8,10 @@
     "moduleResolution": "node",
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
-    "target": "es2015",
+    "allowUnreachableCode": false,
```
Comment (RasmusKjeldgaard) [inline tsconfig.base.json:11]: remove

File: tsconfig.base.json [hunk 9]
```diff
@@ -8,9 +8,10 @@
...
+    "target": "es2017",
     "typeRoots": ["node_modules/@types"],
-    "lib": ["es2017", "dom", "es6", "es2015.iterable"],
+    "lib": ["es2017", "dom"],
```
Comment (RasmusKjeldgaard) [inline tsconfig.base.json:14]: What is happening here?

File: libs/designsystem/src/lib/components/angular-component-lib/utils.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import { fromEvent } from 'rxjs';
+
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/src/lib/components/angular-component-lib/utils.ts:2]: Not related to specific line: 
Seems like the path for this (angular-component-lib) is not configurable. It is autogenerated by stencil, based on where the proxies.ts file is outputted.

File: tsconfig.base.json (hunk 7 above)
Comment (RasmusKjeldgaard) [reply tsconfig.base.json:12]: The way I see it, this change is sane enough. 

This has no consequences for the angular library, as the tsconfig for that sets the target to `es2017`. 
So instead, we could do as suggested here and set the base config to `es2017`, and remove it from the angular specific config. 

Furthermore, the cookbook has a browserslistrc that angular uses to determine the needed JS output. Running `npx browserslist`, I see the lowest supported versions as:

```
chrome 90
edge 89
firefox 78
ios_saf 13.0-13.1
safari 13
```
Which is all well within [support for es2017](https://caniuse.com/?search=es2017).

File: tsconfig.base.json (hunk 9 above)
Comment (RasmusKjeldgaard) [reply tsconfig.base.json:14]: The target (previously `es2015`) is supplemented by the `lib` option which is used to indicate what extra JS API’s/features the compiled project uses on top of `es2015`. So if we supply polyfills for some or all `es2017`-specific features we can indicate that in the `lib` config. 

But I can only think of use cases where we want to compile to a lower target like `es2015` to support older browsers, but then supply polyfills for some features, so the runtime can be es2017. And to silence the compiler we then specify what functionality we ensure ourselves to deliver through the `lib` config. 

Since our previous target was `es2015` (which is also es6), there should never have been a need for these, as we already have a target that supports these features. 

In conclusion I cant see anything wrong with removing this. It is overwritten for our published package, so it is only relevant for the compiled cookbook code, and the core library.

https://www.typescriptlang.org/tsconfig#lib
https://stackoverflow.com/a/42097465
https://www.meziantou.net/which-version-of-ecmascript-should-i-use-in-the-typescript-configuration.htm

Comment (RasmusKjeldgaard) [reply tsconfig.base.json:14]: We should consider to bump the target of [tsconfig.lib.json] (https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/tsconfig.lib.json) (published angular lib) to `es2017` since we are already 'promising' this and more by setting the 'lib' to `es2018`. To my knowledge we are not supplying polyfills to close that gap.

File: tsconfig.base.json (hunk 7 above)
Comment (RasmusKjeldgaard) [reply tsconfig.base.json:12]: We should consider to bump the target of tsconfig.lib.json (published angular lib) to es2017 since we are already 'promising' this and more by setting the 'lib' to es2018. To my knowledge we are not supplying polyfills to close that gap.

File: package.json [hunk 11]
```diff
@@ -24,7 +24,7 @@
   "scripts": {
     "ng": "nx",
     "nx": "nx",
-    "start": "ng serve",
+    "start": "ng serve && npm run build:core:watch",
```
Comment (RasmusKjeldgaard) [inline package.json:27]: This is not going to work - ng serve does not terminate, so core is never built. 

Consider concurrently, npm-run-all or similar to run the processes besides each other.

**Comment by RasmusKjeldgaard** [comment]
TODO:
- [ ] Find an alternative to themeColor directive, as it is angular specific
- [x] Create separate branch for autogenerating mock based on this branch, and merge to this
- [ ] Create separate branch for publishing (#1630) and merge to this 
- [ ] Create both npm packages locally and install in DRB as test
- [ ] **!IMPORTANT!** Before merge, make sure to verify that moved helpers and style files have any new changes from master.

File: libs/core/src/components/badge/badge.component.tsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { Component, ComponentInterface, h, Host, Prop } from '@stencil/core';
+
+import { ColorHelper, ThemeColor } from '../../helpers';
```
Comment (RasmusKjeldgaard) [inline libs/core/src/components/badge/badge.component.tsx:3]: Remove unused ```ColorHelper```- and set up eslint?

**Comment by RasmusKjeldgaard** [comment]
Consider updating to [Stencil v.2.9.0](https://github.com/ionic-team/stencil/releases/tag/v2.9.0) before merging.

**Review by MadsBuchmann — CHANGES_REQUESTED** [review]

File: apps/cookbook/src/app/examples/icon-example/icon-example.component.scss [hunk 13]
```diff
@@ -1,4 +1,4 @@
-@import './libs/designsystem/src/lib/scss/utils';
+@import '~@kirbydesign/core/src/scss/utils';
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply apps/cookbook/src/app/examples/icon-example/icon-example.component.scss:1]: That is true. Looking at my suggestion again I think we should perhaps just ignore it? I have trouble seeing exactly what value it would add. It would also make it even more rigid to make changes to configuration/packaging, which I do not think is desirable.

File: package.json [hunk 14]
```diff
@@ -24,16 +24,17 @@
   "scripts": {
     "ng": "nx",
     "nx": "nx",
-    "start": "ng serve",
+    "start": "npm-run-all --parallel \"ng serve\" \"build:core:watch\"",
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply package.json:27]: Wooohooo!

File: .github/workflows/publish.yml [hunk 15]
```diff
@@ -5,18 +5,19 @@ on:
...
-    name: Bump and publish new package version to npm
+  # If core has changes, we want to bump both core and designsystem
+  # If only designsystem has changes, we want to just bump designsystem
+  bump_package_versions: 
+    name: Bump version and publish changed packages
     runs-on: ubuntu-latest
-    if: github.event.pull_request.merged
     steps:
-      - name: Checkout master branch
+      - name: Checkout repo
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply .github/workflows/publish.yml:19]: Good point! 

We already have a issue on it here actually now that I turn on my brain: https://github.com/kirbydesign/designsystem/issues/1518 

I would LOVE to carry that one out.

File: .gitignore [hunk 16]
```diff
@@ -9,6 +9,14 @@
 *.styles.ts
 /libs/designsystem/src/lib/polyfills/*.min.js
 
+# compiled core library output
+/libs/core/loader
+/libs/core/dist/
+/libs/core/scss/
+/libs/core/custom-elements.d.ts
+/libs/core/custom-elements.json
+/libs/core/src/components.d.ts
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply .gitignore:18]: cool 😄

File: .prettierignore [hunk 17]
```diff
@@ -1,2 +1,11 @@
...
 /.github/
+/libs/designsystem/src/lib/components/web-component-proxies.component.ts
+/libs/designsystem/src/lib/components/angular-component-lib/**.*
+/libs/designsystem/src/lib/components/angular-component-lib/utils.ts
+/libs/core/dist/**
+/libs/core/loader/**
+/libs/core/component-docs.d.ts
+/libs/core/component-docs.json
+/libs/core/src/components.d.ts
+readme.md
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply .prettierignore]: Ah cool :D!

File: libs/core/package.json [hunk 18]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "@kirbydesign/core",
+  "version": "0.0.1",
+  "description": "Core web-components for the Kirby Design System.",
+  "module": "dist/index.js",
+  "main": "dist/index.cjs.js",
+  "types": "dist/types/index.d.ts",
+  "collection": "dist/collection/collection-manifest.json",
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/core/package.json:8]: cool 🦖

File: libs/designsystem/testing-base/src/lib/directives/mock.theme-color.directive.ts [hunk 19]
```diff
@@ -5,7 +5,6 @@ import { ThemeColor, ThemeColorDirective } from '@kirbydesign/designsystem';
 @Directive({
   // tslint:disable-next-line:directive-selector
   selector: `kirby-card[themeColor],
-             kirby-badge[themeColor],
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/designsystem/testing-base/src/lib/directives/mock.theme-color.directive.ts:8]: Ah I misunderstood the selector in the first pass. Thank you for the clarification.

File: libs/designsystem/src/lib/app-initialize.ts [hunk 20]
```diff
@@ -0,0 +1,19 @@
+import { DOCUMENT } from '@angular/common';
+import { APP_INITIALIZER } from '@angular/core';
+
+import { defineCustomElements } from '@kirbydesign/core/loader';
+
+export const appInitialize = (doc: Document) => {
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/app-initialize.ts]: Cool 😄

File: libs/designsystem/src/lib/components/avatar/avatar.component.scss [hunk 21]
```diff
@@ -21,7 +21,7 @@ $badge-diameter: $avatar-badge-size;
   --kirby-badge-position: absolute;
   --kirby-badge-right: #{get-badge-position($diameter-small)};
   --kirby-badge-top: #{get-badge-position($diameter-small)};
-  --kirby-badge-zindex: 2;
+  --kirby-badge-z-index: 2;
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/avatar/avatar.component.scss]: Sounds good. The z-index maps seems to be pretty... case specific?... 

So adding 'avatarBadge' does not seem to be poor form to me.

File: libs/designsystem/tsconfig.lib.json [hunk 22]
```diff
@@ -4,6 +4,7 @@
     "outDir": "../../dist/out-tsc",
     "target": "es2015",
     "declaration": true,
+    "inlineSourceMap": true,
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/designsystem/tsconfig.lib.json]: How did it go 😄?

File: libs/designsystem/src/lib/animation/kirby-animation.ts [hunk 23]
```diff
@@ -1,12 +1,12 @@
...
-    SHORT = parseInt(styles.transitionDurations.short),
-    LONG = parseInt(styles.transitionDurations.long),
-    EXTRA_LONG = parseInt(styles.transitionDurations.extraLong),
+    QUICK = parseInt(DesignTokenHelper.transitionDuration('quick')),
+    SHORT = parseInt(DesignTokenHelper.transitionDuration('short')),
+    LONG = parseInt(DesignTokenHelper.transitionDuration('long')),
+    EXTRA_LONG = parseInt(DesignTokenHelper.transitionDuration('extraLong')),
   }
-  export const Easing = styles.transitionEasings;
+  export const Easing = DesignTokenHelper.transitionEasings();
```
[the remark this answers is not in the document]
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/animation/kirby-animation.ts:11]: As is is just totally fine!

**Review by MadsBuchmann — DISMISSED** [review]
It's an honor to finally be able to write these four letters with accompanying gif: 

LGTM!
Happy Coder

File: libs/designsystem/tsconfig.lib.json (hunk 22 above)
[the remark this answers is not in the document]
Comment (RasmusKjeldgaard) [reply libs/designsystem/tsconfig.lib.json]: We have a silent compiler 🤐

**Review by MadsBuchmann — APPROVED** [review]
